libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,560 @@
+
+	// Remove anything from the stack top that can be freely generated or dupped from deeper on the stack.
+	while (!stack.empty())
+	{
+		if (canBeFreelyGenerated(stack.back()))
+			stack.pop_back();
+		else if (auto offset = util::findOffset(stack | ranges::views::reverse | ranges::views::drop(1), stack.back()))
+		{
+			if (*offset + 2 < 16)
+				stack.pop_back();
+			else
+				break;
+		}
+		else
+			break;
+	}
+
+	// TODO: there may be a better criterion than overall stack size.
+	if (stack.size() > 12)
+		// Deduplicate and remove slots that can be freely generated.
+		stack = compressStack(move(stack));
+	return stack;
+}
+
+Stack StackLayoutGenerator::compressStack(Stack _stack)
+{
+	optional<size_t> firstDupOffset;
+	do
+	{
+		if (firstDupOffset)
+		{
+			if (_stack.size() - *firstDupOffset - 1 > 1)
+				std::swap(_stack.at(*firstDupOffset + 1), _stack.back());
+			std::swap(_stack.at(*firstDupOffset), _stack.back());
+			_stack.pop_back();
+			firstDupOffset.reset();
+		}
+		for (auto&& [offset, slot]: _stack | ranges::views::enumerate)
+			if (canBeFreelyGenerated(slot) || util::findOffset(_stack | ranges::views::take(offset), slot))
+				firstDupOffset = offset;
+	}
+	while (firstDupOffset);
+	return _stack;
+}
+
+Stack StackLayoutGenerator::propagateStackThroughBlock(Stack _exitStack, CFG::BasicBlock const& _block)
+{
+	Stack stack = std::move(_exitStack);
+	for (auto& operation: _block.operations | ranges::views::reverse)
+		stack = propagateStackThroughOperation(stack, operation);
+	return stack;
+}
+
+list<pair<CFG::BasicBlock const*, CFG::BasicBlock const*>> StackLayoutGenerator::collectBackwardsJumps(CFG::BasicBlock const& _entry) const
+{
+	list<pair<CFG::BasicBlock const*, CFG::BasicBlock const*>> backwardsJumps;
+	util::BreadthFirstSearch<CFG::BasicBlock const*>{{&_entry}}.run([&](CFG::BasicBlock const* _block, auto _addChild) {
+		std::visit(util::GenericVisitor{
+			[&](CFG::BasicBlock::MainExit const&) {},
+			[&](CFG::BasicBlock::Jump const& _jump)
+			{
+				if (_jump.backwards)
+					backwardsJumps.emplace_back(_block, _jump.target);
+				_addChild(_jump.target);
+			},
+			[&](CFG::BasicBlock::ConditionalJump const& _conditionalJump)
+			{
+				_addChild(_conditionalJump.zero);
+				_addChild(_conditionalJump.nonZero);
+			},
+			[&](CFG::BasicBlock::FunctionReturn const&) {},
+			[&](CFG::BasicBlock::Terminated const&) {},
+		}, _block->exit);
+	});
+	return backwardsJumps;
+}
+
+optional<Stack> StackLayoutGenerator::getExitLayoutOrStageDependencies(
+	CFG::BasicBlock const& _block,
+	set<CFG::BasicBlock const*> const& _blocksWithExitLayout,
+	list<CFG::BasicBlock const*>& _toVisit
+) const
+{
+	return std::visit(util::GenericVisitor{
+		[&](CFG::BasicBlock::MainExit const&) -> std::optional<Stack>
+		{
+			return Stack{};
+		},
+		[&](CFG::BasicBlock::Jump const& _jump) -> std::optional<Stack>
+		{
+			if (_jump.backwards)
+			{
+				if (auto* info = util::valueOrNullptr(m_layout.blockInfos, _jump.target))
```

Comment:
Why do you use `m_layout.blockInfos` here but `_blocksWithExitLayout` below?

Reply:
Can the two cases be unified?

Reply:
For backwards jumps I just use whatever the last pass gave me as entry layout for the jump target (even though it may not yet be final) - that's stored in ``m_layout.blockInfos`` - and don't ever stage anything for revisit. I could actually probably just return ``m_layout.blockInfos[_jump.target].entryLayout`` directly, but that might implicitly create an empty one - which is probably fine, but a bit weird.

For forward jumps, I want to visit the target first, so I check if I already have an exit layout for it, resp. if already have visited it in the current pass (maybe calling ``_blocksWithExitLayout`` should remain to be called ``_visited`` after all...) and only then return the layout of the block and otherwise just stage the jump target for visit.